c2rust-analyze/src/c_void_casts.rs

```diff
@@ -320,41 +405,44 @@ impl<'tcx> CVoidCasts<'tcx> {
             };
             let func_ty = func.ty(&body.local_decls, tcx);
 
-            // For [`CVoidCastDirection::From`], we only count
-            // a cast from `*c_void` to an arbitrary type in the subsequent block,
-            // searching forward.
-            let find_first_cast_succ_block = |get_cast| {
-                let successor_block_id = target.unwrap();
-                body.basic_blocks()[successor_block_id]
-                    .statements
-                    .iter()
-                    .find_map(get_cast)
-            };
-
-            // For [`CVoidCastDirection::From`], we only count
-            // a cast to `*c_void` from an arbitrary type in the same block,
-            // searching backwards.
-            let find_last_cast_curr_block =
-                |get_cast| bb_data.statements.iter().rev().find_map(get_cast);
-
             for direction in CVoidCastDirection::from_callee(ty_callee(tcx, func_ty))
                 .iter()
                 .copied()
             {
                 use CVoidCastDirection::*;
-                let c_void_ptr = match direction {
+                let c_void_ptr_place = match direction {
                     From => destination,
                     To => args[0].place().unwrap(),
                 };
-                let c_void_ptr = CVoidPtr::checked(c_void_ptr, &body.local_decls, tcx);
+                let c_void_ptr = CVoidPtr::checked(c_void_ptr_place, &body.local_decls, tcx);
                 let get_cast =
-                    move |stmt: &Statement<'tcx>| c_void_ptr.get_cast_from_stmt(direction, stmt);
+                    |stmt: &Statement<'tcx>| c_void_ptr.get_cast_from_stmt(direction, stmt);
                 let cast = match direction {
-                    From => find_first_cast_succ_block(get_cast),
-                    To => find_last_cast_curr_block(get_cast),
+                    From => Self::find_first_cast_succ_block(
+                        get_cast,
+                        &body.basic_blocks()[target.unwrap()].statements,
+                    ),
+                    To => Self::find_last_cast_curr_block(
+                        get_cast,
+                        &bb_data.statements,
+                        &c_void_ptr_place,
+                    ),
                 };
-                if let Some(cast) = cast {
-                    self.insert(direction, cast);
+                if let Some((statement_index, cast)) = cast {
+                    let loc = Location {
+                        statement_index,
+                        block: match direction {
+                            To => BasicBlock::from_usize(block),
+                            From => target.unwrap(),
+                        },
+                    };
+                    self.insert(loc, direction, cast.clone());
+
+                    let loc = Location {
+                        statement_index: bb_data.statements.len(),
```

Comment:
What does it mean to have an out-of-bounds `statement_index` for a `Location`?

Reply:
I think a function like this in `util.rs` would make things a lot clearer.  I'd've thought it'd be in `rustc` already.

```rust
pub fn terminator_location(block: BasicBlock, block_data: &BasicBlockData) -> Location {
    Location {
        block,
        statement_index: block_data.statements.len(),
    }
}
```

Reply:
a7ddc515